### PR TITLE
When re-creating listeners in EVMConnect/EthConnect/FabConnect use the last event block

### PR DESCRIPTION
**Pull request by peterbroadhurst**
See #1531 

Every time Core starts, it confirms all the listeners exist that it needs for all of:
- Contract listeners registered by applications
- `BatchPin` listeners for any multi-party namespaces
- Token listeners - indirectly via `/activatepool`
    - Not changed in this PR - see #1531 comments for justification

This is needed for two cases:
1. In case the listeners were deleted operationally for any reason (intentionally or otherwise) - such as a DR scenario
2. In the v1.2->v1.3 migration, where we intentionally orphan all the core-wide listeners in favor of namespace-specific ones

In both of these cases, the logic was naive for the `newest`/`latest` case - where it treats "now" as the latest event.

That is a problem, as it might be a long time since the original listener was created, and it might have delivered many events already. In which case there might be "gaps" we skip over between the last event detected by the old listener, and the first event we detect on the new one.

This PR closes that window, by looking at the `protocolId` of the last event - which is architecturally required to be a lexicographically sorted string identifying the position of the event on the blockchain.

If there's been a previous event, that's used by the blockchain connector as the `fromBlock` for the new listener.

Also, if `"oldest"` or an explicit block number is used, and the block number is before the block determined by the above process, then the newer block is used.

### CODE REVIEW

File: internal/multiparty/manager.go [hunk 1]
```diff
@@ -155,7 +155,23 @@ func (mm *multipartyManager) configureContractCommon(ctx context.Context, migrat
...
 	}
 
-	subID, err := mm.blockchain.AddFireflySubscription(ctx, mm.namespace, current)
+	// For the case that we're establishing a listener from "latest" we obtain the protocol ID
+	// of the latest event confirmed from the blockchain for a given subscription.
+	// This protocolID should be parsed and used by the blockchain plugin if SubOptsFirstEventNewest
+	// is passed through, and the listener does not exist.
+	fb := database.BlockchainEventQueryFactory.NewFilter(ctx).Sort("-protocolid").Limit(1)
+	latestEvents, _, err := mm.database.GetBlockchainEvents(ctx, mm.namespace.Name, fb.Eq(
+		"listener", nil,
```
Comment (EnriqueL8) [inline internal/multiparty/manager.go:164]: Why are we not using the multiparty listener ID here? We don't have access to it?

Comment (peterbroadhurst) [reply internal/multiparty/manager.go:164]: Specifically the thing that distinguishes the `blockchainevents` for the multi-party `BatchPin` events, is the lack of a `listener` reference (there isn't one in the FF model for the `BatchPin` case).

**Comment by peterbroadhurst** [comment]
PR has been updated based on the conversation here, so that in the case a specific block number is supplied the recreated listener will still restart on the block before the last event (if that's after the specified block): https://github.com/hyperledger/firefly/issues/1531#issuecomment-2187054888

**Comment by EnriqueL8** [comment]
Will be testing this one today e2e

File: internal/multiparty/manager.go (hunk 1 above)
Comment (EnriqueL8) [reply internal/multiparty/manager.go:164]: Ah makes sense

**Review by EnriqueL8 — COMMENTED** [review]
Re-reviewing I thought about a crazy edge case... The protocol ID is block number/transaction number /event number so if we recreate a listener with block number of the last protocol ID but we haven't consumed all the events from that block could we lose those events? Is it safer to go one down of the procotol id or am I over complicating this? Would have to see how FireFly acknowledges events from FFTM and maybe the assumption you have made is that you get all the events per block in the webscoket?

**Comment by EnriqueL8** [comment]
Have tested this one E2E by doing the following experiment:
- Setup a 1.3.0 FF stack
- Deploy a contract and a listener with the option to listen from newest
- Delete the listener from EVMConnect
- Restart the FF with the code from this PR
- Find the new listener created in EVMConnect and make sure the `fromBlock` matches the last protocol ID for the last event - 1

**Review by EnriqueL8 — APPROVED** [review]
This looks good to me!

**Comment by peterbroadhurst** [comment]
> we recreate a listener with block number of the last protocol ID but we haven't consumed all the events from that block could we lose those events? Is it safer to go one down of the procotol id or am I over complicating this?

This is the reason we go one block back:
https://github.com/hyperledger/firefly/blob/8dd5a0d2899a0b002d080a60f15ab64770ef1be0/internal/blockchain/ethereum/eventstream.go#L216-L218

**Comment by EnriqueL8** [comment]
Yeah I realised with https://github.com/hyperledger/firefly/pull/1534#issuecomment-2205567325

**Pull request merged**